### PR TITLE
add e2e tests

### PR DESCRIPTION
**Pull request by veezhang**

### CODE REVIEW

File: tests/e2e/suites.go [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+	}
+	helmInstall(helmName, helmNamespace, workloadNamespaces, helmArgs...)
+}
+
+func setupKruise() {
+	if !e2econfig.TestConfig.InstallKruise {
+		return
+	}
+	helmName := "kruise"
+	helmNamespace := "default"
```
Comment (MegaByte875) [inline tests/e2e/suites.go:143]: set namespace to kruise-system

File: tests/e2e/suites.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+	if err != nil {
+		if !strings.Contains(err.Error(), "release: not found") {
+			framework.ExpectNoError(err, "failed to helm status %s%s", namespace, name)
+		}
+		return false
+	}
+	return true
+}
+
+func helmInstall(name, namespace string, workloadNamespaces []string, args ...string) {
```
Comment (MegaByte875) [inline tests/e2e/suites.go:199]: I think install helm charts [kruise,cert-manager,nebula-operator] by e2e.sh shell scripts  more concise,  we can check if they installed or not in e2e framework

File: tests/e2e/nebulacluster/nebulacluster.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+					if tc.StoragedReplicas < 3 {
+						replicaFactor = 1
+					}
+					executeSchema := fmt.Sprintf(
+						"CREATE SPACE IF NOT EXISTS e2e_test(partition_num=%d,replica_factor=%d);",
+						5*tc.StoragedReplicas, replicaFactor) +
+						"USE e2e_test;" +
+						"CREATE TAG IF NOT EXISTS person(name string, age int);" +
+						"CREATE EDGE IF NOT EXISTS like(likeness double);"
+					err = waitForExecuteNebulaSchema(30*time.Second, 2*time.Second, graphLocalAddress, graphLocalPort, "user", "pass", executeSchema)
```
Comment (MegaByte875) [inline tests/e2e/nebulacluster/nebulacluster.go:147]: If we run the e2e test without creating a new kind cluster, the graphLocalAddress and graphLocalPort will not work

**Review by MegaByte875 — APPROVED** [review]

File: tests/e2e/nebulacluster/util.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: v1alpha1.NebulaClusterSpec{
+			Graphd: &v1alpha1.GraphdSpec{
+				PodSpec: v1alpha1.PodSpec{
+					Replicas: pointer.Int32Ptr(1),
+					Image:    "vesoft/nebula-graphd",
+					Version:  "v2-nightly",
```
[the remark this answers is not in the document]
Comment (veezhang) [reply tests/e2e/nebulacluster/util.go:57]: default: `v2-nightly`, and I'll add config for the image.

File: hack/e2e.sh [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+echo "KIND_CONFIG: ${KIND_CONFIG}"
+
+echo "KUBECONFIG: ${KUBECONFIG}"
+echo "DELETE_NAMESPACE: ${DELETE_NAMESPACE}"
+echo "INSTALL_KUBERNETES: ${INSTALL_KUBERNETES}"
+echo "DELETE_NAMESPACE_ON_FAILURE: ${DELETE_NAMESPACE_ON_FAILURE}"
+
+ginkgo ./tests/e2e \
+  -- \
+  --install-kubernetes=${INSTALL_KUBERNETES} \
```
[the remark this answers is not in the document]
Comment (veezhang) [reply hack/e2e.sh:49]: goland shell scripts plugins

File: hack/e2e.sh [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+INSTALL_KUBERNETES=${INSTALL_KUBERNETES:-true}
+UNINSTALL_KUBERNETES=${UNINSTALL_KUBERNETES:-true}
+INSTALL_CERT_MANAGER=${INSTALL_CERT_MANAGER:-true}
+INSTALL_KRUISE=${INSTALL_KRUISE:-true}
+INSTALL_NEBULA_OPERATOR=${INSTALL_NEBULA_OPERATOR:-true}
+KIND_NAME=${KIND_NAME:-e2e-test}
+KIND_CONFIG=${KIND_CONFIG:-${ROOT}/hack/kind-config.yaml}
+
+KUBECONFIG=${KUBECONFIG:-~/.kube/${KIND_NAME}.kind.config}
```
[the remark this answers is not in the document]
Comment (veezhang) [reply hack/e2e.sh:29]: Oh, you will be able to use the existing kubernetes via the following command:
```shell
INSTALL_KUBERNETES=false make e2e
```

File: tests/e2e/nebulacluster/util.go [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+	nebula "github.com/vesoft-inc/nebula-go"
+	"github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1"
+	"github.com/vesoft-inc/nebula-operator/pkg/label"
+	e2eutil "github.com/vesoft-inc/nebula-operator/tests/e2e/util"
+)
+
+func getNebulaCluster(namespace, name string) *v1alpha1.NebulaCluster {
+	imagePullPolicy := corev1.PullIfNotPresent
+	storageClassName := "standard" // standard, fast-disks
```
[the remark this answers is not in the document]
Comment (veezhang) [reply tests/e2e/nebulacluster/util.go:46]: The following priority will be applied:

1.  `STORAGE_CLASS` environment variable
2. The default  ClassName in kubernetes
3. The other ClassName in kubernetes

File: tests/e2e/nebulacluster/util.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (veezhang) [reply tests/e2e/nebulacluster/util.go:57]: Default to use `v2-nightly`, and you can set the `NEBULA_VERSION` environment variable to specify the version.
```shell
NEBULA_VERSION=v2.0.1 make e2e
```

File: tests/e2e/suites.go [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+	helmName := "cert-manager" // nolint: goconst
+	helmNamespace := "cert-manager"
+	workloadNamespaces := []string{helmNamespace}
+	helmArgs := []string{
+		"install", helmName,
+		"cert-manager",
+		"--repo", "https://charts.jetstack.io",
+		"--namespace", helmNamespace,
+		"--create-namespace",
+		"--version", "v1.3.1",
```
[the remark this answers is not in the document]
Comment (veezhang) [reply tests/e2e/suites.go:132]: Default to use `v1.3.1`, and you can set the `INSTALL_CERT_MANAGER_VERSION` environment variable to specify the version.
```shell
INSTALL_CERT_MANAGER_VERSION=v1.3.1 make e2e
```

File: tests/e2e/suites.go [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+	if !e2econfig.TestConfig.InstallKruise {
+		return
+	}
+	helmName := "kruise"
+	helmNamespace := "default"
+	workloadNamespaces := []string{"kruise-system"}
+	helmArgs := []string{
+		"install", helmName,
+		"https://github.com/openkruise/kruise/releases/download/v0.8.1/kruise-chart.tgz",
+	}
```
[the remark this answers is not in the document]
Comment (veezhang) [reply tests/e2e/suites.go:148]: Default to use `v0.8.1`, and you can set the `INSTALL_KRUISE_VERSION` environment variable to specify the version.
```shell
INSTALL_KRUISE_VERSION=v0.8.1 make e2e
```

File: tests/e2e/suites.go (hunk 1 above)
Comment (veezhang) [reply tests/e2e/suites.go:143]: The `helmNamespace` means the helm's namespace nor the crd's namespace to install.

In the helm repo,  there have a namespace resource `kruise-system`. So, if set `helmNamespace` to `kruise-system`, we need to create `kruise-system` first or add `--create-namespce` to helm command.
But it already a namespace `kruise-system` in helm repo.
There's a conflict. I haven't found a better way at the moment.

File: tests/e2e/suites.go (hunk 2 above)
Comment (veezhang) [reply tests/e2e/suites.go:199]: The shell script should be executed before the test starts. And there may be different behaviors between each test case. With `go`, we can have better control.

File: tests/e2e/nebulacluster/nebulacluster.go (hunk 3 above)
Comment (veezhang) [reply tests/e2e/nebulacluster/nebulacluster.go:147]: It dose work. And it have tested.
The network traffic forwards to `api-server`, and then to `graphd`. 
We can access the `api-server`, so the network traffic that forward to `graphd`.

**Review by MegaByte875 — APPROVED** [review]

**Comment by MegaByte875** [comment]
LGTM

**Pull request merged**